### PR TITLE
Trash feature

### LINKED ISSUE #330: Trash feature
Implement reusable trash feature like in TEXTman, LEADman, and Foliolabs DOCman

### PR DESCRIPTION
**Pull request by jebbdomingo**
closes #330

### CODE REVIEW

File: code/libraries/joomlatools/library/controller/behavior/trashable.php [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+     * Trash entities
+     *
+     * @param  KControllerContextInterfaceInterface $context
+     * @return void
+     */
+    protected function _actionTrash(KControllerContextInterface $context)
+    {
+        $data             = $context->request->data;
+        $data->trashed    = 1;
+        $data->trashed_on = gmdate('Y-m-d H:i:s');
```
Comment (ercanozkaya) [inline code/libraries/joomlatools/library/controller/behavior/trashable.php]: @jebbdomingo @johanjanssens shall we add trashed_by here?

File: code/libraries/joomlatools/library/controller/toolbar/trashable.php [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+class KControllerToolbarTrashable extends KControllerToolbarDecorator
+{
+    protected function _afterBrowse(KControllerContextInterface $context)
+    {
+        $controller = $this->getController();
+
+        if ($controller->getRequest()->getQuery()->get('trashed', 'int')) {
+            $this->addCommand('restore', array('allowed' => $controller->canEdit()));
+        } else {
+            $this->addCommand('trash', array('allowed' => $controller->canEdit()));
```
Comment (ercanozkaya) [inline code/libraries/joomlatools/library/controller/toolbar/trashable.php]: @johanjanssens @jebbdomingo should we not remove the delete command here? Also canDelete and canAdd sounds like a better permission pair to use here to me.

File: code/libraries/joomlatools/library/controller/toolbar/trashable.php [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        $query      = $this->getController()->getRequest()->getQuery();
+
+        $command->label = $translator->translate('Restore');
+        $command->append(array(
+            'attribs' => array(
+                'data-action' => 'restore',
+            )
+        ));
+
+        $command->label = 'Restore from trash';
```
Comment (ercanozkaya) [inline code/libraries/joomlatools/library/controller/toolbar/trashable.php]: @jebbdomingo you are providing the label twice here, and the second is not translated

File: code/libraries/joomlatools/library/controller/toolbar/trashable.php (hunk 2 above)
Comment (jebbdomingo) [reply code/libraries/joomlatools/library/controller/toolbar/trashable.php]: @ercanozkaya @johanjanssens Though there are suggested practices on handling soft delete, my idea is to provide the user the option if they want to hard delete immediately. Let me know your thoughts.

Comment (ercanozkaya) [reply code/libraries/joomlatools/library/controller/toolbar/trashable.php]: @jebbdomingo If you show trash and delete together you cannot easily convey the difference between them. I think we should hide one

Comment (johanjanssens) [reply code/libraries/joomlatools/library/controller/toolbar/trashable.php]: @jebbdomingo Doing both doesn't make sense it's too complex for a user to get. If you implement a trashing solution ot's always a two step process to do the hard delete. First you do the soft delete and then you do the hard delete. 

Users are used to this flow, both from Joomla and from their operating system. If we change that they will get very confused.

File: code/libraries/joomlatools/library/controller/behavior/trashable.php (hunk 1 above)
Comment (johanjanssens) [reply code/libraries/joomlatools/library/controller/behavior/trashable.php]: @ercanozkaya @jebbdomingo Yes, we  do that everywhere, so best to do a trashed_by too, this is then useful for LOGman to render activities and include who did it.

Comment (jebbdomingo) [reply code/libraries/joomlatools/library/controller/behavior/trashable.php]: @ercanozkaya @jebbdomingo Agreed, I added it already.

**Pull request merged**